### ISSUE TITLE
[0.9.5] query performance

### ISSUE THREAD

**Issue by esecules** [influxdata/influxdb#4979]
I just upgraded from 0.9.3 and I am seeing very high query latency (> 10s) on queries that used to take less than 1 second. Under a load of purely writes the system is much happier.  

For example, in my setup I have a wrapper around each write operation that counts how long the write took and enters that as a point as  "value". The only tag is the serial number of the source of the wrapped data. This hasn't given me problems with having too many series in the past. 

my query is 
```
SELECT mean("value") FROM "influx.latency.write" WHERE time > now() - 6h GROUP BY time(1m)
``` 

heres my config file too:
```
reporting-disabled = true
[[udp]]
enabled = false
[[graphite]]
enabled = false
[http]
https-certificate = "/etc/ssl/influxdb.pem"
https-enabled = false
pprof-enabled = false
enabled = true
bind-address = ":8086"
write-tracing = false
log-enabled = true
auth-enabled = false
[monitor]
store-interval = "10s"
store-enabled = true
store-database = "_internal"
[hinted-handoff]
retry-rate-limit = 0
enabled = true
max-age = "168h"
max-size = 1073741824
dir = "/srv/hh"
purge-interval = "1h"
retry-interval = "1s"
retry-max-interval = "1m"
[admin]
https-certificate = "/etc/ssl/influxdb.pem"
https-enabled = false
enabled = true
bind-address = ":8083"
[collectd]
enabled = false
[cluster]
shard-writer-timeout = "10s"
write-timeout = "5s"
[meta]
election-timeout = "1s"
leader-lease-timeout = "500ms"
hostname = "localhost"
bind-address = ":8088"
heartbeat-timeout = "1s"
cluster-tracing = false
retention-autocreate = true
commit-timeout = "50ms"
dir = "/srv/meta"
raft-promotion-enabled = true
[opentsdb]
enabled = false
[shard-precreation]
check-interval = "10m"
advance-period = "30m"
enabled = true
[data]
wal-enable-logging = true
wal-dir = "/var/lib/influxdb/wal"
max-wal-size = 104857600
wal-flush-interval = "10m"
dir = "/srv/db"
wal-partition-flush-delay = "2s"
[continuous_queries]
recompute-previous-n = 2
enabled = true
compute-runs-per-interval = 10
compute-no-more-than = "2m"
log-enabled = true
recompute-no-older-than = "10m"
[retention]
check-interval = "30m"
enabled = true
```

**Comment by beckettsean** [comment]
@esecules how many points are in that time range? E.g. what's the return of:

`SELECT count("value") FROM "influx.latency.write" WHERE time > now() - 6h`

**Comment by esecules** [comment]
the result right now is 27117 points. There is a pretty constant _influx_ of data into this measurement so its also going to be representative of when I was experiencing query troubles.

**Comment by beckettsean** [comment]
That is a trivial number of points to query, and should return in tens or maybe hundreds of milliseconds. I am unable to come up with a reason it would take so long, other than resource contention.

0.9.5 uses the `bz1` engine by default, which is more RAM intensive than the `b1` engine that was the default with 0.9.3. Is your RAM saturated?

**Comment by esecules** [comment]
I am giving the container 6GiB to work with, and it gets up to about 90-95% usage under queries like this.

**Comment by zstyblik** [comment]
Hello,

I'm possibly seeing the same issue with 0.9.6.1 and TSM1.

> SELECT mean(value) FROM telegraf."default".disk_inodes_free WHERE host = 'builder' AND time > 402767h AND time < now() GROUP BY time(1m);

takes about 2.1minutes to return.

> SELECT count(value) FROM telegraf."default".disk_inodes_free WHERE host = 'builder' AND time > 402767h AND time < now();

yields 419253. This is instantaneous. When `GROUP BY time(1m)` is added, whether it makes sense or not, it takes ~ 2+ minutes to finish as in previous case.

InfluxDB is running in Docker container in/at VM with 4 CPU cores and 6GB RAM. top shows CPU@200%, RAM ~ 12%.

**Comment by zstyblik** [comment]
It seems that slow-down really is in `GROUP BY`. `SELECT mean(value)` or even `SELECT value` is what you'd call instantaneous(fetched ~ 600k points). @beckettsean , is there anything else that can be done to debug this issue?

**Comment by zstyblik** [comment]
... However, if you further narrow down the query, then it gets faster =~ 8s.

> SELECT mean(value) FROM telegraf."default".disk_inodes_free WHERE host = 'builder' AND path = '/' AND time > 402767h AND time < now() GROUP BY time(1m);

I also ran similar query against ~ 20k data points and it was similarly fast. Still, 600k data points isn't that much of data, is it? I mean, one metric per one host in 10s intervals should total to 3.1M data points per year.

**Comment by beckettsean** [comment]
@zstyblik thanks for the additional details. There's not really a bug here so much as a performance issue with many GROUP BY buckets.

The larger ~500k point query is creating about 4000 buckets and doing a mean for each. That's non-trivial, compared to calculating the single mean for all 500k points. I don't think it should be two orders of magnitude slower, however. 

@benbjohnson more fodder for the query engine rewrite. Looks like the bucketing code is very inefficient in comparison to simply ripping points off disk.

**Comment by jwilder** [comment]
Should be fixed in the `0.11` release.
